### PR TITLE
feat(learn): Add/Remove Tests for Imperial-Metric Converter Project

### PR DESCRIPTION
**Pull request by SaintPeter**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] All the files I changed are in the same world language, for example: only English changes, or only Chinese changes, etc.

~~* Removes the two "Information Security" tests as per #39435~~
* Add additional tests for invalid numbers/units/both number and units and final output check
* Improve error reporting for non-XHR errors.  If an assertion fails the error is in the property `message`, so I have ORed the `message` for all tests. This provides a better experience for debugging.

## Testing
I have tested from GitPod against my personal project (which passes) and against the boilerplate (which fails).

## Note on 'invalid *' tests
The current example project does not pass some of these tests.  As noted in #39470, it has some issues with some invalid units.  It also returns the error as `{ error: "invalid unit" }`.  If this is the standard, then the project/boilerplate will need to be updated to state as such.  For the moment I have adopted test strategy of checking for a value in the `.error` property and plain-text return value.
IE: 
```js
data.error === 'invalid number' || data === 'invalid number'
```

### CODE REVIEW

File: curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/metric-imperial-converter.english.md
Comment (nhcarrigan) [inline curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/metric-imperial-converter.english.md:28]: Can we add the validation for the example URL? Like on the exercise tracker, to ensure campers cannot submit the example project and pass. Something like this:

```yaml
  - text: I can provide my own project, not the example URL.
    testString: "getUserInput => {
      assert(!(new RegExp('.*/metric-imperial-converter--freecodecamp\\.repl\\.co')).test(getUserInput('url')));
    }
    "
```

**Review by naomi-lgbt — COMMENTED** [review]
Just one quick question. 

Also, I'll keep an eye on this as if it gets merged first I'll need to rebase #39613

File: curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/metric-imperial-converter.english.md
Comment (SaintPeter) [reply curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/metric-imperial-converter.english.md:28]: Yeah, I can do that.

Comment (SaintPeter) [reply curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/metric-imperial-converter.english.md:33]: Done.  New commit added.

**Comment by SaintPeter** [comment]
Merge conflict should be resolved.

**Review by naomi-lgbt — DISMISSED** [review]
image
image

As mentioned, the example project does fail the one test. The remainder of the tests pass locally against the example, and all tests pass against @SaintPeter's project.

**Comment by SaintPeter** [comment]
If you would like to change against my project:  
https://metric-imperial.pearlygatesoftware.com/

File: curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/metric-imperial-converter.english.md
Comment (Sky020) [inline curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/metric-imperial-converter.english.md:102]:
```suggestion
  - text: If my number is invalid, returned will be <code>'invalid number'</code>.
```

File: curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/metric-imperial-converter.english.md
Comment (Sky020) [inline curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/metric-imperial-converter.english.md:111]:
```suggestion
  - text: If both are invalid, return will be <code>'invalid number and unit'</code>.
```

File: curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/metric-imperial-converter.english.md
Comment (Sky020) [inline curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/metric-imperial-converter.english.md:144]:
```suggestion
  - text: My return will consist of the <code>initNum</code>, <code>initUnit</code>, <code>returnNum</code>, <code>returnUnit</code>, and <code>string</code> spelling out units in format <code>'{initNum} {initial_Units} converts to {returnNum} {return_Units}'</code> with the result rounded to 5 decimals.
```

**Review by ShaunSHamilton — CHANGES_REQUESTED** [review]
Just some formatting of the texts.

**Comment by Sky020** [comment]
Thanks for this, Rex.

Where are we on the `L` vs `l` return units? Are we going to instruct all lowercase return values?

Pending that, I see this as _ready-to-go_.

**Comment by SaintPeter** [comment]
I'm not sure that there is a "consensus", but there was some discussion on the related thread at #39617.

**Comment by SaintPeter** [comment]
I have made the changes as proposed.

These can be tested against my project, here:
https://metric-imperial.pearlygatesoftware.com/

I don't understand all of the other files showing as changed.  I may have mucked up the Git history.  If so, I can try to fix or I can resubmit as needed.

File: curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/metric-imperial-converter.english.md
Comment (nhcarrigan) [inline curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/metric-imperial-converter.english.md:92]:
```suggestion
  - text: All incoming units should be accepted in both upper and lower case, but should be returned in both the <code>initUnit</code> and <code>returnUnit</code> in lower case, except for liter, which should be represented as an uppercase <code>'L'</code>.
```

I would wrap this in code tags like the other mentions of units.

File: curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/metric-imperial-converter.english.md
Comment (nhcarrigan) [inline curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/metric-imperial-converter.english.md:130]:
```suggestion
  - text: If both the unit and number are invalid, returned will be <code>'invalid number and unit'</code>.
```

Verbiage tweak here.

File: curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/metric-imperial-converter.english.md
Comment (nhcarrigan) [inline curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/metric-imperial-converter.english.md:171]: Just a side thought here - in the interest of internationalisation, do we want to explicitly require the American spelling here? Other English dialects will potentially use `kilometres` instead.

**Review by naomi-lgbt — CHANGES_REQUESTED** [review]
A couple of nitpicks. 

I'm not sure what happened to the commit history here - I tried tracing it back but can't seem to find the cause. This *may* need to be rebased.

File: curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/metric-imperial-converter.english.md
Comment (RandellDawson) [reply curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/metric-imperial-converter.english.md:171]: Always use the American spelling.

**Comment by SaintPeter** [comment]
Holy Toledo! I managed to rebase it properly to remove the spurious commits.  I also accepted @nhcarrigan's suggestions.

**Comment by moT01** [comment]
I rebased this again quick since it had conflicts.

**Comment by moT01** [comment]
Do you want to make a PR to the [demo-projects repo](https://github.com/freeCodeCamp/demo-projects) with your example project @SaintPeter? You should remove the user stories from the landing page so it looks [something like this.](https://metric-imperial-converter.freecodecamp.rocks/)

File: curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/metric-imperial-converter.md
Comment (moT01) [inline curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/metric-imperial-converter.md:163]:
```suggestion
  - text: My return will consist of the <code>initNum</code>, <code>initUnit</code>, <code>returnNum</code>, <code>returnUnit</code>, and <code>string</code> spelling out units in the format <code>'{initNum} {initial_Units} converts to {returnNum} {return_Units}'</code> with the result rounded to 5 decimals.
```

File: curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/metric-imperial-converter.md
Comment (moT01) [inline curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/metric-imperial-converter.md:144]:
```suggestion
        const data1 = await $.get(getUserInput('url') + '/api/convert?input=mi');
        assert.approximately(data1.initNum, 1, 0.001);
        assert.approximately(data1.returnNum, 1.60934, 0.001);
```

One test for "if nothing is provided it will default to 1"

**Review by moT01 — COMMENTED** [review]
I made a few small suggestions, they probably aren't even needed - but I would add them. Other than that, this LGTM 🎉 

I think we are fine to merge this when it's ready and, once it hits production, we can merge the PR to the boilerplate. The example project can probably be updated at any point.

**Review by moT01 — APPROVED** [review]
LGTM 🎉

**Comment by raisedadead** [comment]
Any reason we are not merging this yet, other than a second review @moT01 ?

**Comment by moT01** [comment]
Nope @raisedadead

**Comment by RandellDawson** [comment]
I will review it now.

**Review by RandellDawson — CHANGES_REQUESTED** [review]
@SaintPeter For some reason, I thought I had already requested changes on this PR, but I do not see them.  I tested everything against your created project and everything works fine, but we need the example project in the demo-projects repo updated so that all the tests on `/learn` (except the first).  Currently, the example project does not pass all the tests.  Would you mind updating that too?

Once this PR is merged, then we would still need another PR to have this project conform to the specifications laid out in issue https://github.com/freeCodeCamp/freeCodeCamp/issues/39721  (i.e. the use of "You" instead of "I".

We do not want to merge this PR until the boilerplate and the example project are approved and read to be merged, so that all we need to do is wait till the changes of this PR hit production.  The only missing part is the example project.  The boilerplate has been approved already.

**Comment by SaintPeter** [comment]
@RandellDawson Amazingly, I guess I don't have a PR open for the example project.  I'll see about spinning mine up.

**Review by RandellDawson — APPROVED** [review]

**Comment by RandellDawson** [comment]
@SaintPeter I have tested everything against the example project (PR https://github.com/freeCodeCamp/demo-projects/pull/37) and everything passes.  In a separate PR, someone else can work on fulfilling the requirements referenced in issue https://github.com/freeCodeCamp/freeCodeCamp/issues/39721.

Again, thanks for your hard work on all 3 of the PRs related to this project.

@nhcarrigan @Sky020 You show outstanding change requests, so if you are good with the changes, please approve and merge this PR.  Thanks.

**Review by naomi-lgbt — DISMISSED** [review]
My changes have been addressed - I'm good with approving this. I'll give @Sky020 a chance to look as well before merge. 

LGTM~! 🥳

File: curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/metric-imperial-converter.md
Comment (Sky020) [inline curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/metric-imperial-converter.md:171]:
```suggestion
        assert.equal(data.string, '2 mi converts to 3.21868 km')
```

**Review by ShaunSHamilton — CHANGES_REQUESTED** [review]
I am not sure about this, but I am not aware of the requirement for the return to be full words `km vs kilometer`

I would prefer `km` anyway, because it could be `kilometre` too...

**Comment by nhcarrigan** [comment]
For what it's worth, that requirement was previously established and is present in the example project:

image

**Comment by Sky020** [comment]
@nhcarrigan Thanks for that. I missed it in my project.

Then, we at least need to change this section of the user-story/test:
> spelling out units in the format <code>'{initNum} {initial_Units} converts to {returnNum} {return_Units}'</code>

File: curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/metric-imperial-converter.md
Comment (Sky020) [inline curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/metric-imperial-converter.md:171]: To clarify the issue I am seeing here:
- `initial_units` and `return_units` are not referenced before this
- I missed the HTML example output (when I did this project), and followed this pattern:
```
{initNum} {initUnit} converts to {returnNum} {returnUnit}
```

I see two options:
1) Change the tests and HTML example output to not need the _Full-English_ unit name
2) Make the challenge more difficult with this user story as is, but be clear (somewhere else) that there is a requirement to return the _Full-English_ unit names...at some point

Comment (nhcarrigan) [reply curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/metric-imperial-converter.md:171]:
```yml
  - text: My return will consist of the <code>initNum</code>, <code>initUnit</code>, <code>returnNum</code>, <code>returnUnit</code>, and <code>string</code>. <code>string</code> will take the format <code>'{initNum} {initUnit} converts to {returnNum} {returnUnit}'</code> with the converted result rounded to 5 decimals, and <code>initUnit</code> and <code>returnUnit</code> values should be spelled out (e.g. `km` becomes `kilometer`).
```
I do lean more toward keeping the requirement - how is this?

Comment (Sky020) [reply curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/metric-imperial-converter.md:171]: @nhcarrigan Almost. I would not have `initUnit` and `returnUnit` be the same (namespace) as the others. Maybe even something more verbose: `initUnitString & returnUnitString`

Comment (nhcarrigan) [reply curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/metric-imperial-converter.md:171]: That's a good approach, I think.

**Comment by SaintPeter** [comment]
This is my last outstanding PR (of 3).  Are we waiting on some specific changes here?

**Comment by nhcarrigan** [comment]
I believe the changes in @Sky020's review discussion are all that is left.

**Comment by RandellDawson** [comment]
It seems overly complex to have the user provide the shorthand version of a unit (like `mi` for `miles`) and expect the final string property to return the long version of the two (`miles` for `mi` and `kilometers` for `km`), but I am fine as long as user stories are clear about it and the tests correctly test what is expected.  If we are worried about some kind of backwards compatibility, then the user stories/tests should allow for both versions.

**Comment by SaintPeter** [comment]
@RandellDawson I don't think that this is the appropriate venue for this discussion. These requirements were in the original project and were unchanged by my additions.  

The purpose of this PR was to implement the missing tests for this project.  Along the way I added some clarification and made slight changes to the user stories and tests to be more consistent.  That's about it.  

If you would like to propose additional changes to this project, I suggest we open an issue to discuss it.  Right now there are students who are missing out on the benefit of having tests on `/learn` until this PR is merged.

**Comment by moT01** [comment]
I created a new issue to add/discuss the changes that are still needed. I think we should merge this as-is since the issues remaining were preexisting.

**Comment by RandellDawson** [comment]
@Sky020 Are you fine with merging this PR, since @moT01 has created an issue outlining the concerns?

**Review by ShaunSHamilton — APPROVED** [review]
This LGTM 👍 . We can definitely go ahead with smaller PRs, I think 😅 

@SaintPeter Rex, Thanks for these changes.

**Pull request merged**
